Python3 and sugargame port

srevinsaju opened the pull request:
I returned with my abandoned project,
working compared to #10

srevinsaju commented:
Fixes #9

quozl commented:
Tested 1e17b78 on Ubuntu 19.04; black screen, only the toolbar shown.

```log
1577688670.054241 WARNING root: Activity bundle /usr/share/sugar/activities/falabracman.activity does not specify a license
/usr/share/sugar/activities/falabracman.activity/falabracman.py:20: PyGIWarning: Gtk was imported without specifying a version first. Use gi.require_version('Gtk', '3.0') before import to ensure that the right version gets loaded.
  from gi.repository import Gtk
Traceback (most recent call last):
  File "/usr/share/sugar/activities/falabracman.activity/falabracman.py", line 621, in run
    ((self.screen_width-SCREEN_WIDTH)/2, 0), (SCREEN_WIDTH, ALTURA_BARRA))
ValueError: subsurface rectangle outside surface area
1577688673.474658 ERROR root: Event queue full!
1577688673.481792 ERROR root: Event queue full!
1577688673.490686 ERROR root: Event queue full!
...
Traceback (most recent call last):
  File "/usr/share/sugar/activities/falabracman.activity/sugargame/event.py", line 115, in _visibility_cb
    self.update_display()
  File "/usr/share/sugar/activities/falabracman.activity/sugargame/event.py", line 98, in update_display
    pygame.event.post(pygame.event.Event(pygame.VIDEOEXPOSE))
pygame.error: Event queue full
...
1577688676.314474 ERROR root: Event queue full!
1577688676.322523 ERROR root: Event queue full!
1577688676.330830 ERROR root: Event queue full!
1577688676.337456 ERROR root: Event queue full!
Traceback (most recent call last):
... etc
```

srevinsaju commented:
which version of pygame?  
@quozl , please check. if the previous version works for you before testing this. if the previous doesn't, you might have missed some dependencies

_Sent from my Galaxy Note2 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

Hrishi1999 commented:
Tested https://github.com/sugarlabs/falabracman-activity/commit/1e17b785f2d7ee9bf4ca1475458fc9f012c83dfe, it loads fine to the loading screen. 
Language selection menu opens. When you press Play, nothing happens. Traceback here:

> Traceback (most recent call last):
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 647, in run
>     self.game_instance()
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 574, in game_instance
>     self.status = Status(self, dic)
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 326, in __init__
>     self.avanzarNivel()
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 342, in avanzarNivel
>     self.dic, self.grossini, self.groupsinni)
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 268, in __init__
>     self.nuevaPalabra(grossini)
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 274, in nuevaPalabra
>     self.parent, self.palabra, self.groupsinni, self.grupoLagos)
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 237, in __init__
>     lettersprite_obj.spritefx([groupsinni, lagos, self], letra)
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 194, in spritefx
>     self.image = self.parent.dameLetra(self.parent.letrasEncendidas, letra)
>   File "/home/hrishi/falabracman-activity/falabracman.py", line 537, in dameLetra
>     return dictLetras.get(chr(letra), dictLetras["*"])
> TypeError: an integer is required
> 1577690334.164979 ERROR root: Event queue full!
> 1577690334.174012 ERROR root: Event queue full!
> 1577690334.183183 ERROR root: Event queue full!
> 1577690334.190032 ERROR root: Event queue full!
> 1577690334.200257 ERROR root: Event queue full!
> 1577690334.207542 ERROR root: Event queue full!
> 1577690334.215519 ERROR root: Event queue full!

srevinsaju commented:
>return dictLetras.get(chr(letra), dictLetras["*"])
TypeError: an integer is required

I used chr to convert the integer to string, for me the message was different
```
key error: 82
(because 82 was not found)
```

try changing the encoding on the top of the file to utf-8 and let me know. thanks for testing @Hrishi1999 and @Quozl

srevinsaju commented:
@Quozl error confirmed, I got the error while testing on sugar iso.

Hrishi1999 commented:
@srevinsaju Mentioning utf-8 didn't work.

srevinsaju commented:
Previously Falabracman was forced to run at a fixed resolution. With commit https://github.com/sugarlabs/falabracman-activity/pull/11/commits/25ede2ec4217d4e0a2340ae8b60effdcd54f177c and onwards, the items have been automatically resized corresponding to the screen size.

srevinsaju commented:
@Hrishi1999 I wasn't able to reproduce your error. but I fixed quozl 's.
wrt IRC, try removing  the only chr() on line 545.
there is no reason why I kept it there. thanks

This might help, this was the argument passed to it :
```
{'A': <Surface(56x84x32 SW)>, 'B': <Surface(55x84x32 SW)>, 'C': <Surface(53x84x32 SW)>, 'D': <Surface(60x84x32 SW)>, 'E': <Surface(50x84x32 SW)>, 'F': <Surface(50x84x32 SW)>, 'G': <Surface(59x84x32 SW)>, 'H': <Surface(61x84x32 SW)>, 'I': <Surface(28x84x32 SW)>, 'J': <Surface(32x84x32 SW)>, 'K': <Surface(59x84x32 SW)>, 'L': <Surface(47x84x32 SW)>, 'M': <Surface(72x84x32 SW)>, 'N': <Surface(61x84x32 SW)>, 'O': <Surface(62x84x32 SW)>, 'P': <Surface(53x84x32 SW)>, 'Q': <Surface(62x84x32 SW)>, 'R': <Surface(56x84x32 SW)>, 'S': <Surface(52x84x32 SW)>, 'T': <Surface(50x84x32 SW)>, 'U': <Surface(59x84x32 SW)>, 'V': <Surface(56x84x32 SW)>, 'W': <Surface(79x84x32 SW)>, 'X': <Surface(56x84x32 SW)>, 'Y': <Surface(55x84x32 SW)>, 'Z': <Surface(53x84x32 SW)>, '*': <Surface(39x84x32 SW)>
```

Check in your python interpreter if :  
```
>>> dict1 = {'A':1, 'B':2, 'C':3}
>>> dict1.get('A', '*') # Should return 1
>>> dict1.get('B', '*') # Should return 2
>>> dict1.get('D', '*') # Should return '*'
```

Hrishi1999 commented:
Tested https://github.com/sugarlabs/falabracman-activity/pull/11/commits/e5d4ddf2c27acad7fa159ef566fe19b0847c389f. The activity works if I remove `chr`.

srevinsaju commented:
> Tested [e5d4ddf](https://github.com/sugarlabs/falabracman-activity/commit/e5d4ddf2c27acad7fa159ef566fe19b0847c389f). The activity works if I remove `chr`.

It works on my sytsem with integers, works on your system with chr.
The only solution is a vague workaround.

srevinsaju commented:
@Hrishi1999 fixed it on https://github.com/sugarlabs/falabracman-activity/pull/11/commits/dbf9d980e3558390237988d96b7567542d3a2512

Hrishi1999 commented on falabracman.py:
```diff
@@ -192,7 +196,7 @@ def __init__(self, parent):
 
     def spritefx(self, grossini, letra):
-
+        print(self.parent.letrasEncendidas)
```
Remove this.

srevinsaju commented:
@chimosky @quozl please finalize review this activity. @Hrishi1999 has reviewed it already. Thanks

walterbender commented:
Port seems to work pretty well.

srevinsaju commented:
Thanks for reviewing @walterbender . I was eagerly waiting for a reviewer :P

quozl commented:
Fixes https://github.com/sugarlabs/falabracman-activity/issues/9

Fixes https://github.com/sugarlabs/falabracman-activity/issues/3

chimosky commented:
@quozl kindly test and review as I can't at the moment.

quozl commented:
Thanks.  Reviewed.  Tested.

* [ ] frame rate is quite high; consumes a lot of CPU and power on my test systems, consider if the frame rate is correct for purpose,
* [ ] cursor is invisible; is it needed?
* [ ] a PyGIWarning for Gtk version,
* [ ] returning from options to main menu gives two sound clicks instead of one,
* [ ] no way to escape credits; have to wait for a long time,
* [ ] collision detection with the ponds of water seems unreliable or inaccurate.

These aren't really Port to Python 3 issues, as they likely were present beforehand?

srevinsaju commented:
> * [ ]  frame rate is quite high; consumes a lot of CPU and power on my test systems, consider if the frame rate is correct for purpose,
* frame rate: well haven't really done anything about that on the port. the one before porting had the same issue. pygame warns the user not to load images in delays as had been done already, but we can't omit the splash screen, or the credits.
These are some of the frame rates:
```
falabracman.py:        clock = pygame.time.Clock()
falabracman.py:        clock.tick(4)
falabracman.py:        clock = pygame.time.Clock()
falabracman.py:            clock.tick(10)
falabracman.py:        clock = pygame.time.Clock()
falabracman.py:            clock.tick(20)
```
I think reducing it beyond this will create screen tears but idk,

> * [ ]  cursor is invisible; is it needed?

* cursor: cursor was forced to be hidden for some reason by the originals creators. @chimosky had previously tested it and have confirmed this game is a keyboard-only-game. 

> * [x]  a PyGIWarning for Gtk version,

Fixed on https://github.com/sugarlabs/falabracman-activity/pull/11/commits/873a0cf068d64c47eb3539335dcaa2eade50db3b

> * [ ]  returning from options to main menu gives two sound clicks instead of one,

cant reproduce it :(
I created an assumed fix on https://github.com/sugarlabs/falabracman-activity/pull/11/commits/1e40d7b26e1b6aaa3714cec795bef7cac7422148 
@quozl please check if the double clicks are still there 

> * [ ]  no way to escape credits; have to wait for a long time,


* the credits : the time set previously was 10s for splash and 10s for the credits thingy, so I have reduced it to 4s and 3s. well the only option is to add an escape key to stop it, but it won't work because of the loop, it will work for the time between the first delay (splash) and the second delay(credits), but this too may not be intuitive for kids.


> * [ ]  collision detection with the ponds of water seems unreliable or inaccurate.

* collision is based on pygame's collision object. it depends on the png image provide d as the ponds in the picture, as long as the ponds aren't purely rectangle, there will be some inaccuracies, as png's are rendered by pygame as rectangle. 

> These aren't really Port to Python 3 issues, as they likely were present beforehand?

walterbender commented:
I agree with the analysis above:

* The cursor was set invisible by the creators and is not used. It might be nice to have instructions (or help in the help activity) but that is beyond the scope of the port to Python 3.
* The collision does seem off sometimes, but that too is a function of the original code. Also, sometimes the game starts in a condition where it is impossible to advance because all paths lead to a pond. Again, outside of the scope of the port.

Another unrelated consideration: the "You lose" screen is a bitmap and only available in Portuguese at the moment. At some point, it would be good to make it possible to localize.

srevinsaju commented:
* Yes @walterbender, the algorithm used in the code sometimes places the pond on the initial position of the falabracman, but this is completely random. The creators of the game hadn't kept a constraint on the positions of the pond, but yes, it beyond port to python3
* The you lose or you win is in bitmap, which I had a plan of making an only english bitmap version, and i thought it would be ok if i raise the PR after the GCI heat gets off.
* In addition to Python3 and sugargame port, what I achieved:
- translated the whole code which was spanish or portugese to English to make it intuitive to new programmers who wish to contribute.
- Added english words and added language english to the dictionary and converted the portugese words to english to make it more user friendly, to kids who are not spanish or portugese
- Added the english option in Options and make it the default language instead of `bra`
PS: I am stuck, with this activity on the task list, and i have many of them merged too
